lib/Runtime/Library/InJavascript/Intl.js

```diff
@@ -508,133 +518,191 @@
-            // `|${ALPHA}{4}`         +                                         //               / 4ALPHA              ; or reserved for future use
-            // `|${ALPHA}{5,8}`       +                                         //               / 5*8ALPHA            ; or registered language subtag
-            `|${ALPHA}{4,8}`          +                                         //              ~/ 4*8ALPHA            ; (paraphrased: combined previous two lines)
-            ')\\b';
+        if (parts[4]) {
+            ret.variants = parts[4];
+            // leading "-" is already in parts[4]
+            ret.base += parts[4];
+        }
 
-        // Use matching groups only when needed
-        const LANG_TAG_BASE = `\\b(${language})\\b`         +                   // langtag       = language
-                              `\\b(?:-(${script}))?\\b`     +                   //                 ["-" script]
-                              `\\b(?:-(${region}))?\\b`     ;                   //                 ["-" region]
-        const LANG_TAG_EXT  = `\\b((?:-${variant})*)\\b`    +                   //                 *("-" variant)
-                              `\\b((?:-${extension})*)\\b`  +                   //                 *("-" extension)
-                              `\\b(?:-${privateuse})?\\b`   ;                   //                 ["-" privateuse]
-        const langtag       = `\\b${LANG_TAG_BASE}\\b${LANG_TAG_EXT}\\b`;
+        if (parts[5]) {
+            ret.extensions = parts[5];
 
-        const LANG_TAG      = `\\b(?:${langtag}|${privateuse}|${grandfathered})\\b`;  // Language-Tag  = ...
+            // parse the extension to find the unicode (-u) extension
+            const extensionParts = _.split(parts[5], "-");
+            for (let unicodeExtensionStart = 0; unicodeExtensionStart < extensionParts.length; ++unicodeExtensionStart) {
+                if (extensionParts[unicodeExtensionStart] !== "u") {
+                    continue;
+                }
 
-        // Use ^ and $ to enforce that the entire input string is a langtag
-        const LANG_TAG_BASE_RE    = new RegExp(`^${LANG_TAG_BASE}$`, 'i'); // [1] language; [2] script; [3] region;
-        const LANG_TAG_EXT_RE     = new RegExp(`^${LANG_TAG_EXT}$`,  'i'); //                                       [1] variants; [2] extensions;
-        const LANG_TAG_RE         = new RegExp(`^${LANG_TAG}$`,      'i'); // [1] language; [2] script; [3] region; [4] variants; [5] extensions;
+                let unicodeExtensionsEnd;
+                for (unicodeExtensionsEnd = unicodeExtensionStart + 1; unicodeExtensionsEnd < extensionParts.length && extensionParts[unicodeExtensionsEnd].length > 1; unicodeExtensionsEnd++) {
+                    // do nothing, we just want k to equal the index of the next element whose length is 1
```

Comment:
what are you referring to with `k`? I don't see any variables with that name (which is nice, by the way).

Reply:
oops, renamed the variable without updating the comment.

Reply:
> Maybe we should share some notes across implementations about where it is or isn't possible to reuse ICU's locale processing; c

v8 is about to make a switch to ICU API from BCP 47 based regex. There are two groups of issues:

1. ICU's handling of grandfathered tags and deprecated region/language code is outdated. This is a data issue. I've filed a series of bugs against the ICU. I'm assigned to them and I do have patches (that have been applied to v8's copy of ICU since this spring)

2. I also discovered a couple of bugs and put up a PR against the ICU (it's approved, but not yet merged).  

With the above two issues resolved, v8 will make a switch to ICU with a couple of extra pre/post-processing.

Reply:
I spoke too early. What I wrote above is mainly structure-validation and canonicalization.  Min/max also work fine.

Reply:
Last I had checked V8 (not sure which version) it was still converting unicode attributes to unicode keywords with value = "yes", and converting en-GB-oed to en-GB-x-oed rather than en-GB-oxendict. Chakra does this as well as a direct result of allowing ICU to do the canonicalization, so I assumed V8 used ICU in the same way that we did. In other words, I thought both V8 and Chakra suffered primarily from data issues because of ICU, not structure issues.

Reply:
Thank you for reminding me of 'yes' issue. That's still outstanding.  It's 
https://unicode-org.atlassian.net/browse/ICU-13730 . 

In case of en-GB-oed, it's fixed at least in geCanonicalLocales() along with a number of cases arising from data.  ICU upstream hasn't been fixed, though. [1]

```
var loc1=Intl.getCanonicalLocales("en-GB-oed")
undefined
loc1
["en-GB-oxendict"]
```

Let me check Intl.Locale. 

[1] https://unicode-org.atlassian.net/browse/ICU-13721 
13719, 13720, 13723, 13726 are other bugs about the date update. Perhaps, I'd better consolidate them all into one and make a PR ( the v8/Chromium patch is
https://cs.chromium.org/chromium/src/third_party/icu/patches/locid_map.patch ).

Reply:
Intl.locale is also fixed (as it should be):
```
$ d8 --harmony-locale
d8> new Intl.Locale("en-gb-oed").toString()
"en-GB-oxendict"
```

Reply:
As for the structural validity, [my PR](https://github.com/unicode-org/icu/pull/102)  for https://unicode-org.atlassian.net/browse/ICU-20098 was just merged to the ICU tot.  When I replaced custom regular expressions for BCP 47 structural validity  in v8 with ICU uloc_forLanguageTag/uloc_toLanguageTag with the above PR applied locally, at least there's no regression and one failing test begins to pass.

Reply:
Ah, that's great! The Chromium patch + ICU PR covers all of the cases I tested that ICU didn't handle correctly except for three (plus the -yes issue): 

1. [The UTS35 Likely Subtags algorithm](https://www.unicode.org/reports/tr35/#Likely_Subtags) notes that sh -> sr_Latn and mo -> ro_MD, but in my testing ICU didn't handle that. The UTS page is unclear about where that data comes from -- its not in the subtag registry, and I don't know enough about the CLDR layout to find it there, either.
2. und-Arab-AF maximizes to ar-Arab-AF, which seems wrong -- [Wikipedia](https://en.wikipedia.org/wiki/Languages_of_Afghanistan) says the two primary languages of Afghanistan are Dari (prs) and Pashto (ps), not Arabic (ar). I am not sure if the Arabic script bit of the tag is causing the language to be Arabic, but the above Likely Subtags algorithm says und-Arab-AF should be maximized to fa-Arab-AF (fa == Persian), which also seems more reasonable since at least Dari is a Persian dialect.
3. The UTS35 page also mentions that when the script is Zzzz or the region is ZZ, it should be removed from the tag entirely, but ICU seems to accept it.

I can file ICU issues for any/all three if theyre actually incorrect behavior.